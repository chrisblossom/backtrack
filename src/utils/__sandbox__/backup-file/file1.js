/*
 * file1_modified.js
 */
