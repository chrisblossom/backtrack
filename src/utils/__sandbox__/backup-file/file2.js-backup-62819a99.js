/*
 * file2_modified.js
 */
