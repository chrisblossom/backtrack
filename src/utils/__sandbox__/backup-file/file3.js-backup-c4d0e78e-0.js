/*
 * file3_modified_another_time.js
 */
