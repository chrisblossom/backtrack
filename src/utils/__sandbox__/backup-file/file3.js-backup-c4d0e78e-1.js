/*
 * file3_modified_third_time.js
 */
