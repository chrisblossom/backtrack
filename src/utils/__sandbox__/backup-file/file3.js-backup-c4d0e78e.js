/*
 * file3_modified_again.js
 */
