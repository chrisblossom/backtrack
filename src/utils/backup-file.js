/* @flow */

import path from 'path';
import { existsSync } from 'fs';
import { move } from 'fs-extra';
import { getFileHash } from './get-file-hash';

import { rootPath } from '../config/paths';

type Return = Promise<?{
    file: string,
    filename: string,
}>;

async function backupFile(file: string): Return {
    const hash = getFileHash(file);

    const { base: filename, dir, ext = '' } = path.parse(file);

    const shortHash = hash.slice(0, 8);

    let backupFilename = `${filename}-backup-${shortHash}${ext}`;
    let backupFilePath = path.resolve(dir, backupFilename);

    /**
     * Handle the same file being backed up multiple times
     */
    let previouslyBackedUp = false;
    let backupFileExists = existsSync(backupFilePath);
    if (backupFileExists) {
        let count = 0;
        const maxCount = 10;
        while (
            backupFileExists &&
            previouslyBackedUp === false &&
            count < maxCount
        ) {
            const backupFileHash = getFileHash(backupFilePath);

            /**
             * Check if the existing file has the same hash.
             *
             * Prevents backing up the same file over
             */
            if (backupFileHash === hash) {
                previouslyBackedUp = true;
            }

            backupFilename = `${filename}-backup-${shortHash}-${count}${ext}`;
            backupFilePath = path.resolve(dir, backupFilename);

            backupFileExists = existsSync(backupFilePath);
            count += 1;
        }

        /**
         * Prevent flooding filesystem with backups
         */
        if (count === maxCount && backupFileExists) {
            const relativeFile = path.relative(rootPath, file);

            throw new Error(`Remove all existing backups of: ${relativeFile}`);
        }
    }

    /**
     * Don't backup anything if the file has already been backed up
     */
    if (previouslyBackedUp === false) {
        /**
         * Will fail if backup file already exists
         */
        await move(file, backupFilePath, {
            overwrite: false,
        });

        return {
            file: backupFilePath,
            filename: backupFilename,
        };
    }

    return undefined;
}

export { backupFile };
