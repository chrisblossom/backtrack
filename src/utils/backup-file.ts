import path from 'path';
import { existsSync } from 'fs';
import { move } from 'fs-extra';
import { rootPath } from '../config/paths';
import { getFileHash } from './get-file-hash';

type Return = Promise<
	| {
			file: string;
			filename: string;
	  }
	| undefined
>;

async function backupFile(file: string): Return {
	const hash = getFileHash(file);

	const { base: filename, dir, ext = '' } = path.parse(file);

	const shortHash = hash.slice(0, 8);

	let backupFilename = `${filename}-backup-${shortHash}${ext}`;
	let backupFilePath = path.resolve(dir, backupFilename);

	/**
	 * Handle the same file being backed up multiple times
	 */
	let previouslyBackedUp = false;
	let backupFileExists = existsSync(backupFilePath);
	if (backupFileExists) {
		let count = 0;
		const maxCount = 10;
		while (
			backupFileExists &&
			previouslyBackedUp === false &&
			count < maxCount
		) {
			const backupFileHash = getFileHash(backupFilePath);

			/**
			 * Check if the existing file has the same hash.
			 *
			 * Prevents backing up the same file over
			 */
			if (backupFileHash === hash) {
				previouslyBackedUp = true;
			}

			backupFilename = `${filename}-backup-${shortHash}-${count.toString()}${ext}`;
			backupFilePath = path.resolve(dir, backupFilename);

			backupFileExists = existsSync(backupFilePath);
			count += 1;
		}

		/**
		 * Prevent flooding filesystem with backups
		 */
		if (count === maxCount && backupFileExists) {
			const relativeFile = path.relative(rootPath, file);

			throw new Error(`Remove all existing backups of: ${relativeFile}`);
		}
	}

	/**
	 * Don't backup anything if the file has already been backed up
	 */
	if (previouslyBackedUp === false) {
		/**
		 * Will fail if backup file already exists
		 */
		await move(file, backupFilePath, {
			overwrite: false,
		});

		return {
			file: backupFilePath,
			filename: backupFilename,
		};
	}

	return undefined;
}

export { backupFile };
