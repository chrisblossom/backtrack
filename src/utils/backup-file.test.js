/* @flow */

import path from 'path';

require('./backup-file');

const backupFile = (files) => require('./backup-file').backupFile(files);

describe('backupFile', () => {
    const dir = path.resolve(__dirname, '__sandbox__/backup-file/');
    const cwd = process.cwd();
    let move;

    beforeEach(() => {
        // $FlowIssue
        move = require.requireMock('fs-extra').move;

        jest.mock('fs-extra', () => ({
            move: jest.fn(() => Promise.resolve()),
        }));

        jest.mock('../utils/log', () => ({
            warn: jest.fn(),
        }));

        process.chdir(dir);
    });

    afterEach(() => {
        process.chdir(cwd);
    });

    test('backups file', async () => {
        const file = path.resolve(dir, 'file1.js');

        const result = await backupFile(file);

        expect(result).toMatchSnapshot();
        expect(move.mock.calls).toMatchSnapshot();
    });

    test('backups file inside nested folder', async () => {
        const file = path.resolve(dir, 'nested/nested-file1.js');

        const result = await backupFile(file);

        expect(result).toMatchSnapshot();
        expect(move.mock.calls).toMatchSnapshot();
    });

    test('does nothing if backup file already exists and hash matches', async () => {
        const file = path.resolve(dir, 'file2.js');

        const result = await backupFile(file);

        expect(result).toEqual(undefined);
        expect(move.mock.calls).toEqual([]);
    });

    test('appends -NUM to file when backed up file is modified', async () => {
        const file = path.resolve(dir, 'file3.js');

        const result = await backupFile(file);

        expect(result).toMatchSnapshot();
        expect(move.mock.calls).toMatchSnapshot();
    });

    test('will fail if count >= 10', async () => {
        const file = path.resolve(dir, 'existing/file4.js');

        try {
            expect.hasAssertions();
            await backupFile(file);
        } catch (error) {
            expect(error).toMatchSnapshot();
            expect(move.mock.calls).toEqual([]);
        }
    });
});
