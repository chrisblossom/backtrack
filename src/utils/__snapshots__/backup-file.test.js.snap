// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backupFile appends -NUM to file when backed up file is modified 1`] = `
Object {
  "file": "<PROJECT_ROOT>/file3.js-backup-c4d0e78e-2.js",
  "filename": "file3.js-backup-c4d0e78e-2.js",
}
`;

exports[`backupFile appends -NUM to file when backed up file is modified 2`] = `
Array [
  Array [
    "<PROJECT_ROOT>/file3.js",
    "<PROJECT_ROOT>/file3.js-backup-c4d0e78e-2.js",
    Object {
      "overwrite": false,
    },
  ],
]
`;

exports[`backupFile backups file 1`] = `
Object {
  "file": "<PROJECT_ROOT>/file1.js-backup-7fea248e.js",
  "filename": "file1.js-backup-7fea248e.js",
}
`;

exports[`backupFile backups file 2`] = `
Array [
  Array [
    "<PROJECT_ROOT>/file1.js",
    "<PROJECT_ROOT>/file1.js-backup-7fea248e.js",
    Object {
      "overwrite": false,
    },
  ],
]
`;

exports[`backupFile backups file inside nested folder 1`] = `
Object {
  "file": "<PROJECT_ROOT>/nested/nested-file1.js-backup-c990b9ac.js",
  "filename": "nested-file1.js-backup-c990b9ac.js",
}
`;

exports[`backupFile backups file inside nested folder 2`] = `
Array [
  Array [
    "<PROJECT_ROOT>/nested/nested-file1.js",
    "<PROJECT_ROOT>/nested/nested-file1.js-backup-c990b9ac.js",
    Object {
      "overwrite": false,
    },
  ],
]
`;

exports[`backupFile will fail if count >= 10 1`] = `[Error: Remove all existing backups of: existing/file4.js]`;
